File: java/src/main/java/io/binac/leetcode/ValidParentheses.java
package io.binac.leetcode;

import java.util.Stack;

/**
 * Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
 * <p>
 * <p>An input string is valid if:
 * <p>
 * <p>Open brackets must be closed by the same type of brackets.
 * <p>Open brackets must be closed in the correct order.
 * <p>
 * <p>Note that an empty string is also considered valid.
 * <p>
 * <p>Example 1:
 * <blockquote><pre>
 *     Input: "()"
 *     Output: true
 * </blockquote></pre>
 * Example 2:
 * <blockquote><pre>
 *     Input: "()[]{}"
 *     Output: true
 * </blockquote></pre>
 * Example 3:
 * <blockquote><pre>
 *     Input: "(]"
 *     Output: false
 * </blockquote></pre>
 * Example 4:
 * <blockquote><pre>
 *     Input: "([)]"
 *     Output: false
 * </blockquote></pre>
 * Example 5:
 * <blockquote><pre>
 *     Input: "{[]}"
 *     Output: true
 * </blockquote></pre>
 */
public class ValidParentheses {
    public static class Solution1 {
        public boolean isValid(String s) {
            final int len = s.length();
            Stack<Character> stack = new Stack<>();
            for (int i = 0; i < len; ++i) {
                final char ch = s.charAt(i);
                switch (ch) {
                    case '(':
                        stack.push(ch);
                        break;
                    case ')':
                        if (stack.empty() || stack.pop() != '(')
                            return false;
                        break;
                    case '{':
                        stack.push(ch);
                        break;
                    case '}':
                        if (stack.empty() || stack.pop() != '{')
                            return false;
                        break;
                    case '[':
                        stack.push(ch);
                        break;
                    case ']':
                        if (stack.empty() || stack.pop() != '[')
                            return false;
                        break;
                    default:
                        break;
                }
            }
            return stack.empty();
        }
    }
}

File: python/leetcodepy/valid_parentheses.py
"""
Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

1. Open brackets must be closed by the same type of brackets.
2. Open brackets must be closed in the correct order.

Note that an empty string is also considered valid.

Example 1:

Input: "()"
Output: true

Example 2:

Input: "()[]{}"
Output: true

Example 3:

Input: "(]"
Output: false

Example 4:

Input: "([)]"
Output: false

Example 5:

Input: "{[]}"
Output: true
"""


class Solution1:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        for ch in s:
            if ch == '(' or ch == '[' or ch == '{':
                stack.append(ch)
            else:
                if not stack:
                    return False
                tmp = stack.pop()
                if not (tmp == '(' and ch == ')' or tmp == '[' and ch == ']' or tmp == '{' and ch == '}'):
                    return False
        return not stack
